### PR TITLE
Unpoison the statx buffer for MSAN

### PR DESCRIPTION
**Pull request by gbaraldi**

### CODE REVIEW

File: src/uv-common.c [hunk 1]
```diff
@@ -1002,3 +1002,10 @@ uint64_t uv_metrics_idle_time(uv_loop_t* loop) {
     idle_time += uv_hrtime() - entry_time;
   return idle_time;
 }
+
+void uv__msan_unpoison(const volatile void *a, size_t size) {
```
Comment (Keno) [inline src/uv-common.c]: I think this is better as a `static inline` in the header, so it actually generates no code if msan is disabled.

File: src/uv-common.h [hunk 2]
```diff
@@ -378,6 +378,6 @@ struct uv__loop_internal_fields_s {
 # define _COMPILER_MSAN_ENABLED_
 #endif
 #endif
-void uv__msan_unpoison(const volatile void *a, size_t size);
+static inline void uv__msan_unpoison(const volatile void *a, size_t size);
```
Comment (Keno) [inline src/uv-common.h]: That only works if you put the body into the header also.

Comment (gbaraldi) [reply src/uv-common.h]: Do you think keeping the MSAN define is useful or just move everything inside the function and that's enough?

Comment (Keno) [reply src/uv-common.h]: As is, probably just move it inside the function.

File: src/unix/fs.c [hunk 3]
```diff
@@ -1598,7 +1598,8 @@ static int uv__fs_statx(int fd,
     flags |= AT_SYMLINK_NOFOLLOW;
 
   rc = uv__statx(dirfd, path, flags, mode, &statxbuf);
-
+  uv__msan_unpoison(&statxbuf, sizeof(struct uv__statx));
```
Comment (Keno) [inline src/unix/fs.c]: Shouldn't this go directly into `uv__statx` right where the actual syscall is being done?

File: src/unix/linux-syscalls.c [hunk 4]
```diff
@@ -250,7 +250,9 @@ int uv__statx(int dirfd,
 #if !defined(__NR_statx) || defined(__ANDROID_API__) && __ANDROID_API__ < 30
   return errno = ENOSYS, -1;
 #else
-  return syscall(__NR_statx, dirfd, path, flags, mask, statxbuf);
+  int ret = syscall(__NR_statx, dirfd, path, flags, mask, statxbuf);
+  uv__msan_unpoison(&statxbuf, sizeof(struct uv__statx));
```
Comment (Keno) [inline src/unix/linux-syscalls.c]: Hmm, maybe unpoison on success only? I'm guessing the kernel doesn't touch the buffer on failure.

File: src/unix/fs.c [hunk 5]
```diff
@@ -1619,7 +1619,8 @@ static int uv__fs_statx(int fd,
     uv__store_relaxed(&no_statx, 1);
     return UV_ENOSYS;
   }
-
+  uv__msan_unpoison(&statxbuf, sizeof(struct uv__statx));
```
Comment (Keno) [inline src/unix/fs.c]: I'd still want it inside the uv__statx - just check the error directly in there.

**Comment by gbaraldi** [comment]
Could we merge this? Checking for success inside the function that does the syscall doesn't seem to be how LibUV usually does it.

**Comment by vtjnash** [comment]
Here is how libuv usually handles MSAN (prior to deleting this code and using the glibc wrapper instead), as Keno said: https://github.com/bnoordhuis/libuv/commit/233490819ca4010d0db70a0ae411ae9f73571baa

**Comment by vtjnash** [comment]
c.f. https://github.com/libuv/libuv/pull/311

**Comment by Keno** [comment]
At least I'm consistent ;)

**Comment by bnoordhuis** [comment]
I've opened https://github.com/libuv/libuv/pull/3788 to address this, also for getrandom() ~~(edit: but not yet recvmmsg.)~~

**Comment by gbaraldi** [comment]
Do we backport/pull the upstream changes, or do we wait for an upstream release?
@vtjnash ?

**Pull request closed without merging**